numpyro/infer/einstein/kernels.py

```diff
@@ -205,6 +205,7 @@ def mode(self):
         return self._mode
 
     def compute(self, particles, particle_info, loss_fn):
+        # FIXME(fehiepsi): is the usage of `numpy.random` here safe?
```

Comment:
@OlaRonning Does those `np.random` calls happen 1 time and not happen under jit?

Reply:
@fehiepsi good question, it should happen once even under jit. If I've understood correctly, the return from `npr.randn` should be cached in jaxpr as constants.

Reply:
I see. Thanks, it seems that this is used to have the same random_weights/biases and the same choice of particles
```
particles = particles[npr.choice(particles.shape[0], self.bandwidth_subset)]
```
across vectorized/parallel maps (`vmap`, `pmap`,...). I am not sure about this design decision, especially for the particle subset above. That determination does not match the word "random" as I expected in dropout. Here if we run `jit(compute)` multiple times, I guess the same subset will be selected while if we run `compute` multiple times, different subsets will be selected each time. Could you suggest a note to add to the docs for this behavior?

Reply:
Sorry I've gotten back regarding this. 

I re-read the [SVGD as Moment matching paper](https://arxiv.org/pdf/1810.11693.pdf) and in the Empirical Experiments section (Appendix D), the specify that the random parameters are selected once (at the beginning) and fixed over iteration. That this is consistent with the random feature representation (eq. 5) they use in their analysis is not entirely apparent to me.

The current naming is foremost faithful to the paper that introduces them. However, I appreciate the disconnect with the "common" semantic of an iterative random process. The kernels are written to use JIT compiling, and so a warning that triggers if the kernel is used without JIT compilation would suffice?

Reply:
Regarding the divergence from Jax explicitly parsing the state of the PRNG, I agree that using `numpy.random` is undesirable. With that in mind, I think it should be rewritten using `jax.random`, but it maintains it's current behavior.

Reply:
Thanks for explaining! It is clearer to me now. Unlike other kernels, the particle shapes here are fixed and unlike dropout, the particle subset is also fixed. Looking like we should add an assertation
```python
if self._particles_shape is None:
    self._particles_shape = particles.shape
else:
    assert particles.shape == self._particles_shape, \
        "Expect particles.shape={}, but got {}.".format(self._particles_shape, particles.shape)
```
and revise
```python
if self.bandwidth_subset is not None:
    particles = particles[npr.choice(particles.shape[0], self.bandwidth_subset)]
```
into something like
```python
if self.bandwidth_subset is not None:
    if self._bandwidth_subset_indices is not None:
        self._bandwidth_subset_indices = npr.choice(particles.shape[0], self.bandwidth_subset)
    particles = particles[self._bandwidth_subset_indices
```

This still suffers from the issue: parallel map steinvi with different seeds will use the same weight/bias/subset_indices. To match JAX behavior, I think we can add `rng_key` to the constructor and initialize those infos (weights, bias, subset indices) using that random key. This will require users to specify a random seed for this kernel. What do you think?

Reply:
That looks great! 

Would it be possible to get the `rng_key` using the `primitive` instead of adding it as a parameter? To keep the signature consistent.

Reply:
It seems reasonable but will require running this `compute` method under `seed` handler. Did you run this method under `seed` handler in `svi.init`? We used similar approach in `numpyro.module`:
```python
params = numpyro.param(...)
if params is None:
    rng = numpyro.prng_key()
    params = module.get_initial_params(rng)
```

Reply:
Currently, a kernel object is passed as a parameter when instantiating as in
``` python
stein = Stein(model, WrappedGuide(guide), Adam(1e-3), TraceELBO(),RandomFeatureKernel())
```
and not instantiated by `Stein` (as I remembered it). With that in mind, I think adding a rng_key to `__init__` for `RandomFeatureKernel` is a fine solution
``` python
stein = Stein(model, WrappedGuide(guide), Adam(1e-3), TraceELBO(),RandomFeatureKernel(rng_key))
```

Reply:
I just took a look at your Stein PR #833. It turns out that the kernel is used in `_svgd_loss_and_grads`, which is not run under the `.init()` method. It seems to be tricky to keep the signature consistent here. A solution is to follow SVI/stax/flax/haiku pattern, which is to have a `kernel.init` method that does the initialization job. In SVI or Stein, the `init` method is passed with `rng_key`, so we can use a random key there to initialize the kernel. Something likes
```python
class EinSteinVI(...):
    def init(self, rng_key, ...):
        rng_key, rng_kernel = random.split(rng_key)
        # no-op for all kernels except for `RandomFeatureKernel`
        # (this can also be a chance to precompute any constant terms in each kernel)
        self.kernel_fn.init(rng_key, particles_shape)
```
That way, we can keep the signature consistent. What do you think?

Reply:
I just pushed a commit with the above approach. My only concern is if `particles_shape` is easy to obtain in `Stein.init`. My impression is `Stein.init` and `Stein.update` have a similar signature, so this formation should be available in `init` method. But I might misunderstand something...

Reply:
The stein particles are initialized during `Stein.init`  [here](https://github.com/pyro-ppl/numpyro/blob/c2e6ae08c3de6e4ab4cddc44ce4c47a64e76660e/numpyro/infer/einstein/stein.py#L269-L271), so their shape will be available in `Stein.init`.

Reply:
Hi @OlaRonning, could you take over this PR and make a separate PR for this? I think you understand how to make the change consistent much better than me. It is also faster for us to move Stein inference forward. :)

Reply:
Yes absolutely. Just for clarity, you suggest making a separate PR for `kernel.init` (also updating the `Stein.init`), factor your current `RandomFeatureKernel.init` out of this PR, and merging the remainder?

Reply:
Yup, but you might not need to have several PRs for this: just a single one similar to this but being more completed is enough (e.g. I'm not sure if `def init(self, rng_key, particles_shape):` is a good signature for the init method).

Reply:
I see, thanks for clarifying. I'll code and see how it works out :)